### PR TITLE
Make is_similar behave more consistently

### PR DESCRIPTION
**Pull request by marcelm**
The `is_similar` function used to compare cloneIDs when clustering them within `correct_clone_ids` has a few problems. Many of them were found by @acorbat and are fixed in his version of the function in #36, but instead of adding a fixed function and keeping the one with the problems, this PR fixes the existing function.

- The function is not symmetric: It ignores prefixes or suffixes consisting of `-` only if they occur in the first passed string (`s`), not in the second (`t`).
- It only ignores the `-` characters (it should also ignore `0`).
- String pairs not containing `-` or considered to be similar even when they are shorter than `min_overlap`.
- When the passed strings have different lengths, this did not always raise an exception.
- The `-` characters are only ignored if they are at the beginning or the end, but they can actually be in the middle as well (as in `ACGT----TGCA`) because of the way multiple reads can get  merged into a single molecule.

I could not use @acorbat’s suggested function because it does not take care of the last point.

This PR adds a couple of tests. @acorbat, can you perhaps go through the listed comparisons and check whether the results correspond to what you would expect?

One big downside of this PR is that it makes TREX slower because we can no longer use the `hamming_distance` function from tinyalign (which is implemented in C). `trex run10x` still finishes within 1 minute and a normal-sized dataset, so I think it is acceptable for the moment (runtime was ~20 seconds before). We should focus on correctness first and can then look into improving speed.

### CODE REVIEW

**Comment by acorbat** [comment]
I definitely agree that it's better to have something working and then improve performance when necessary. I would like to write down my thought here just in case ir proves useful. Does it make sense to make something like masks?
```
a = "AC-0TG"
b=  "AC---G"
--------
a_mask = [True, True, False, False, True, True]
b_mask = [True, True, False, False, False, True]

mask = logical_and(a_mask, b_mask)
```

**Comment by marcelm** [comment]
This is essentially how it’s implemented in the PR: The function compares all character pairs. If one or both of the characters are `-` or `0`, the pair is ignored. If they are equal, a match is counted and if they are unequal, a mismatch is counted.

This still has the problem (for which there is the TODO comment) that the number of allowed differences should be reduced if there are `-` or `0`.

**Comment by acorbat** [comment]
I thought that maybe using masks and tinyalign could improve performance. But that is a different question.

With regards to reducing the number of allowed differences when there are ```0``` or ```-```, I would discuss this afterwards. With per cell analysis, I don't expect a big difference, but I haven't quantified it either. I could devise a couple examples for testing and see if we agree on what result we expect. Personally, I am not sure I would add this feature.
If we were to add this, I would have a counter for the number of nucleotides compared and use something like:
```
reduced_threshold = nucleotides_compared * threshold / size_of_barcode
```
If we use masks, we could even do
```
reduced_threshold = sum(mask) * threshold / size_of_barcode
```

Leaving these aside, I think this is a good change. I understand that the default behaviour of the function changes (specially because it now ignores ```0```s as well as ```-```, and missing parts in the middle), so expected out put with default parameter s will change. I could not identify that sequences with ```0```s disappeared. The one thing I would definitely check is that complete sequences (having all 30 nucleotides read) are not corrected into sequences with ```0```s. This could happen because the "correct sequence" is chosen to be the most abundant (addressed in #36 file src/trex/molecule.py line 155).

**Comment by marcelm** [comment]
> I thought that maybe using masks and tinyalign could improve performance. But that is a different question.

Doing anything with the string that requires processing each character individually, such as creating a mask, is slow when done in pure Python:
```
$ python3 -m timeit -s 's="A-GTA-GTACGTACGT"' '[c != "-" for c in s]'
500000 loops, best of 5: 697 nsec per loop
```
At that point, one might as well just do the full Hamming distance computation in pure Python because it’s not that much slower than only creating the mask:
```
$ python3 -m timeit -s 's="ACGTACGTACGTACGT";t="ACGGACGTACGAACGT"' '[1 for (c,d) in zip(s,t) if c != d and c != "-" and d != "-"]'
500000 loops, best of 5: 914 nsec per loop
```
For comparison, Hamming distance in C with tinyalign:
```
$ python3 -m timeit -s 'import tinyalign;s="ACGTACGTACGTACGT";t="ACGGACGTACGAACGT"' 'tinyalign.hamming_distance(s, t)'
2000000 loops, best of 5: 119 nsec per loop
```

> With regards to reducing the number of allowed differences when there are `0` or `-`, I would discuss this afterwards. With per cell analysis, I don't expect a big difference, but I haven't quantified it either.

Ok, let’s take this separately; I will open an issue.

I’ll merge this now as I consider this PR to fix buggy behavior. If some downstream processing needs to change because of this, then that is something we have to deal with separately.

**Review by acorbat — COMMENTED** [review]
I think I never submitted these? I'm really sorry!!!

File: tests/test_similar.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+@pytest.mark.parametrize("s,t,similar", [
+    ("AAAAAAAAAA", "AAAAAAAAAA", True),
+    ("CAAAAAAAAA", "AAAAAAAAAA", True),
+    ("CCAAAAAAAA", "AAAAAAAAAA", False),
+
+    ("", "", False),  # too little overlap
+    ("ACGT", "ACGT", False),  # too little overlap
+    ("TACGT", "TACGT", True),
+    ("TACGT", "TATGT", True),
+    ("TACGT", "TATGT", True),
```
[the remark this answers is not in the document]
Comment (marcelm) [reply tests/test_similar.py:25]: Definitely not.

File: tests/test_similar.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    ("-----TACGT", "-----TACGT", True),
+    ("-----TACGT", "-----TATGT", True),
+    ("-----TACGT", "GGGGGTATGT", True),
+    ("TACGT-----", "TATGT-----", True),
+    ("TACGT-----", "TATGTGGGGG", True),
+    ("TAC-----GT", "TAT-----GT", True),
+    ("TAC-----GT", "TATGGGGGGT", True),
+    ("--TAC---GT--", "--TATGGGGT--", True),
+    ("--TAC---GTGG", "--TATGGGGT--", True),
+    ("GGTAC---GT--", "GGTATGGGGT--", True),
```
[the remark this answers is not in the document]
Comment (marcelm) [reply tests/test_similar.py:38]: The asserts already test for symmetry:
```
assert is_similar(s, t, min_overlap, max_hamming) == similar
# Must be symmetric
assert is_similar(t, s, min_overlap, max_hamming) == similar
```

**Comment by marcelm** [comment]
Opened PR #54 to address review comments.

**Pull request merged**